controllers/dspipeline_params.go

```diff
@@ -174,59 +222,25 @@ func (p *DSPAParams) SetupDBParams(ctx context.Context, dsp *dspa.DataSciencePip
 		p.DBConnection.Port = config.MariaDBHostPort
 		p.DBConnection.Username = p.MariaDB.Username
 		p.DBConnection.DBName = p.MariaDB.DBName
-		if p.MariaDB.PasswordSecret != nil {
-			customCreds = p.MariaDB.PasswordSecret
-		}
-	}
-
-	// Secret where DB credentials reside on cluster
-	var credsSecretName string
-	var credsPasswordKey string
-
-	customCredentialsSpecified := customCreds != nil
-	if customCredentialsSpecified {
-		credsSecretName = customCreds.Name
-		credsPasswordKey = customCreds.Key
-	} else {
-		credsSecretName = p.DBConnection.CredentialsSecret.Name
-		credsPasswordKey = p.DBConnection.CredentialsSecret.Key
-	}
 
-	dbSecret := &v1.Secret{}
-	namespacedName := types.NamespacedName{
-		Name:      credsSecretName,
-		Namespace: p.Namespace,
-	}
-
-	createNewSecret := false
-
-	// Attempt to fetch the specified DB secret
-	err := client.Get(ctx, namespacedName, dbSecret)
-	if err != nil && apierrs.IsNotFound(err) {
-		if !customCredentialsSpecified {
-			generatedPass := passwordGen(12)
-			p.DBConnection.Password = base64.StdEncoding.EncodeToString([]byte(generatedPass))
-			createNewSecret = true
+		// If custom DB Secret provided, use its values.  Otherwise generate a default
+		if p.MariaDB.PasswordSecret != nil {
+			p.DBConnection.CredentialsSecret = p.MariaDB.PasswordSecret
 		} else {
-			log.Error(err, fmt.Sprintf("DB secret [%s] was specified in CR but does not exist.",
-				credsSecretName))
+			p.DBConnection.CredentialsSecret = &dspa.SecretKeyValue{
+				Name: config.DefaultDBSecretNamePrefix + p.Name,
+				Key:  config.DefaultDBSecretKey,
+			}
+		}
+		dbPassword, err := p.RetrieveOrCreateDBSecret(ctx, client, p.DBConnection.CredentialsSecret, log)
```

Comment:
Is the expected behavior for: 

```yaml
spec:
  database:
    mariaDB:
      passwordSecret:
        name: a-custom-db-secret
        key: randomkey
```

that `a-custom-db-secret` is always created? if I provide `a-custom-db-secret` before deploying the dspa, should it not utilize that secret? currently it seems like it will always overwrite this secret

```bash

$ cat dsp.yaml 

apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
kind: DataSciencePipelinesApplication
metadata:
  name: custom-default-secret
spec:
  database:
    mariaDB:
      passwordSecret:
        name: a-custom-db-secret
        key: randomkey
  objectStorage:
    minio:
      image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
  mlpipelineUI:
    image: 'quay.io/opendatahub/odh-ml-pipelines-frontend-container:beta-ui'
---
apiVersion: v1
kind: Secret
metadata:
  name: a-custom-db-secret
stringData:
  randomkey: customDefaultDBSecretTest

$ oc apply -f dspa.yaml 

$ oc get secret a-custom-db-secret -o yaml | yq .data.randomkey | base64 -d  && echo
SJ3Lm6RtY7MB


$


```

Reply:
same for object storage

Reply:
It should *not* overwrite a pre-existing secret (hence the Retrieve in RetrieveOrCreateSecret).  Do you have logs showing when this happens, the code shouldn't allow it

Reply:
This is the code that handles it.  It should only generate a password if we attempt to retrieve the secret and it returns a NotFound error.  any other case either uses the retrieved value or rethrows an error
```go
val, err := p.RetrieveSecret(ctx, client, secretName, secretKey, log)
	if err != nil && apierrs.IsNotFound(err) {
		generatedPass := passwordGen(generatedPasswordLength)
		return base64.StdEncoding.EncodeToString([]byte(generatedPass)), nil
	} else if err != nil {
		log.Error(err, "Unable to create DB secret...")
		return "", err
	}
	log.Info(fmt.Sprintf("Secret [%s] already exists, using stored value.", secretName))
	return val, nil
}
```

Reply:
ah. you are apply these simultaneously.  I wonder if this is a race condition (DSPO detects the secret doesn't exist before the secret was created.  maybe try apply them seperately with a small wait (secret first, then DSPA)

Reply:
ooo good point, I'll try that